### ISSUE TITLE
libp11-0.4.1x: OpenSSL 1.0.2q unable to load pkcs11 engine

### ISSUE THREAD

**Issue by jrgh0** [OpenSC/libp11#307]
Hello,

As this is my first post here, I would like to thank the maintainers of this project.

Going to the matter, Trying to obtain a working embedded system with openssl, PKCS#11 and TPM chip, I've cross-compiled libp11 for ARM.

GNUTLS p11tool is working well with PKCS#11 TPM tokens defined for the TPM 1.2 chip in my embedded hardware, so I can load objects in the tokens, previously defined with opencryptoki, and list them with p11tool.

```
$ p11tool --list-tokens
p11-kit: couldn't find C_GetFunctionList entry point in module: /lib/libp11.so: /lib/libp11.so: undefined symbol: C_GetFunctionList
Token 0:
        URL: pkcs11:model=TPM%20v1.1%20Token;manufacturer=IBM%20Corp.;serial=123;token=ifx
        Label: ifx
        Type: Trust module
        Manufacturer: IBM Corp.
        Model: TPM v1.1 Token
        Serial: 123
        Module: /usr/lib/libopencryptoki.so


$ p11tool --list-all "pkcs11:model=TPM%20v1.1%20Token;manufacturer=IBM%20Corp.;serial=123;token=ifx"
p11-kit: couldn't find C_GetFunctionList entry point in module: /lib/libp11.so: /lib/libp11.so: undefined symbol: C_GetFunctionList
$ p11tool --login --list-all "pkcs11:model=TPM%20v1.1%20Token;manufacturer=IBM%20Corp.;serial=123;token=ifx"
p11-kit: couldn't find C_GetFunctionList entry point in module: /lib/libp11.so: /lib/libp11.so: undefined symbol: C_GetFunctionList
Token 'ifx' with URL 'pkcs11:model=TPM%20v1.1%20Token;manufacturer=IBM%20Corp.;serial=123;token=ifx' requires user PIN
Enter PIN: 
Object 0:
        URL: pkcs11:model=TPM%20v1.1%20Token;manufacturer=IBM%20Corp.;serial=123;token=ifx;id=%15%1f%2d%9b%2d%5f%68%fa%9f%a4%a6%4a%09%d0%ff%1e%8a%42%50%96;object=subCA;type=private
        Type: Private key
        Label: subCA
        Flags: CKA_WRAP/UNWRAP; CKA_PRIVATE; CKA_SENSITIVE; 
        ID: 15:1f:2d:9b:2d:5f:68:fa:9f:a4:a6:4a:09:d0:ff:1e:8a:42:50:96
```

But, as you have seen above libp11 isn't working well and, when I try to use the token objects with openssl pkcs11 engine, I'm in trouble:

```
$ openssl engine pkcs11 -t
(pkcs11) pkcs11 engine
Unable to load module (null)
     [ unavailable ]

$ p11-kit list-modules
p11-kit: couldn't find C_GetFunctionList entry point in module: /lib/libpkcs11.so: /lib/libpkcs11.so: undefined symbol: C_GetFunctionList
```

The files I get after make install, and that I've installed on /lib directory of my filesystem, are next:

```
libp11.so --> /libp11.so.2.5.2   libp11.so.2 --> libp11.so.2.5.2   libp11.so.2.5.2
libpkcs11.so -> pkcs11.so   pkcs11.so
```

I've also created symlink in the engines dir:

```
/usr/lib/engines/libpkcs11.so -> /lib/pkcs11.so
```

As I'm working with p11-kit, the content of /usr/share/p11-kit/modules/pkcs11.module is next:

```
module: /lib/libpkcs11.so
critical: no
trust-policy: yes
managed: yes
log-calls: no
```

I've also tried withouth p11-kit with same result, having engine defined in /etc/ssl/openssl.cnf as you can see below:

```
openssl_conf = openssl_def                                          
                                                                  
[openssl_def]                                                    
engines = engine_section                                            
                                                                  
[engine_section]                                              
pkcs11 = pkcs11_section

[pkcs11_section]                                                    
engine_id=pkcs11                                              
dynamic_path = /lib/libpkcs11.so                        
MODULE_PATH = /lib/pkcs11.so                                        
init = 0
```

I must add that I've tested first with live version of the library (0.4.11), then switched to last released version (0.4.10) with same result.

Following the README.md instructions, I've also tested openssl engine from command line, with next results (see at the awful references to liblibpkcs11.so.so and lib-.so files):

```
$ openssl engine -t dynamic - pre SO_PATH:/lib/libp11.so -pre ID:pkcs11 -pre LIST_ADD:1 -pre LOAD -pre MODULE_PATH:/lib/pkcs11.so
(dynamic) Dynamic engine loading support
[Success]: ID:pkcs11
[Success]: LIST_ADD:1
[Failure]: LOAD
3069269664:error:25066067:DSO support routines:DLFCN_LOAD:could not load the shared library:dso_dlfcn.c:187:filename(liblibpkcs11.so.so): liblibpkcs11.so.so: cannot open shared object file: No such file or directory
3069269664:error:25070067:DSO support routines:DSO_load:could not load the shared library:dso_lib.c:233:
3069269664:error:260B6084:engine routines:DYNAMIC_LOAD:dso not found:eng_dyn.c:467:
[Failure]: MODULE_PATH:/lib/pkcs11.so
3069269664:error:260AC089:engine routines:INT_CTRL_HELPER:invalid cmd name:eng_ctrl.c:131:
3069269664:error:260AB089:engine routines:ENGINE_ctrl_cmd_string:invalid cmd name:eng_ctrl.c:311:
     [ unavailable ]
3069269664:error:25066067:DSO support routines:DLFCN_LOAD:could not load the shared library:dso_dlfcn.c:187:filename(/usr//lib/engines/lib-.so): /usr//lib/engines/lib-.so: cannot open shared object file: No such file or directory
3069269664:error:25070067:DSO support routines:DSO_load:could not load the shared library:dso_lib.c:233:
3069269664:error:260B6084:engine routines:DYNAMIC_LOAD:dso not found:eng_dyn.c:467:
3069269664:error:2606A074:engine routines:ENGINE_by_id:no such engine:eng_list.c:391:id=-
3069269664:error:25066067:DSO support routines:DLFCN_LOAD:could not load the shared library:dso_dlfcn.c:187:filename(/usr//lib/engines/libpre.so): /usr//lib/engines/libpre.so: cannot open shared object file: No such file or directory
3069269664:error:25070067:DSO support routines:DSO_load:could not load the shared library:dso_lib.c:233:
3069269664:error:260B6084:engine routines:DYNAMIC_LOAD:dso not found:eng_dyn.c:467:
3069269664:error:2606A074:engine routines:ENGINE_by_id:no such engine:eng_list.c:391:id=pre
3069269664:error:25066067:DSO support routines:DLFCN_LOAD:could not load the shared library:dso_dlfcn.c:187:filename(/usr//lib/engines/SO_PATH:/lib/libp11.so): /usr//lib/engines/SO_PATH:/lib/libp11.so: cannot open shared object file: No such file or directory
3069269664:error:25070067:DSO support routines:DSO_load:could not load the shared library:dso_lib.c:233:
3069269664:error:260B6084:engine routines:DYNAMIC_LOAD:dso not found:eng_dyn.c:467:
3069269664:error:2606A074:engine routines:ENGINE_by_id:no such engine:eng_list.c:391:id=SO_PATH:/lib/libp11.so
```

So my first question is, where is the origin of this issue? it is related to libp11 or to openssl?

And the second one: there are compatible libp11-0.4.1x and openssl-1.0.2q?

Regards
J.R.

**Comment by dengert** [comment]
The opensc PKCS11 module is opensc-pkcs11.so 
The opensc openssl engine is `libpkcs11.so` or `pkcs11.so` 

You are mixing up the pkcs11.so  with opensc-pkcs11.so 

Change the `MODULE_PATH = /lib/pkcs11.so`  to `MODULE_PATH = /lib/opensc-pkcs11.so`

**Comment by mtrojnar** [comment]
@jrgh0 The OpenSC PKCS#11 module repository is available at https://github.com/OpenSC/OpenSC.  This project (https://github.com/OpenSC/libp11) is an interface for invoking PKCS#11 modules with OpenSSL.

**Comment by dengert** [comment]
There may be some confusion here, including some on my part. 

PKCS\#11 modules have just one function: `C_GetFunctionList` that returns an arrry of function pointers. 

`p11-kit`  has a module that  loads multiple PKCS\#11 modules that are dynamically loadable from some config file. It  then presents them to the caller as one larger module with multiple slots.

`p11tool --list-tokens` says you have one token, and the module used was /usr/lib/libopencryptoki.so`

The OpenSC/libp11 package provides an engine and will load a PKCS\#11 module by default or from the `MODULE_PATH =` in the `openssl.conf`
 
OpenSC has a PKCS\#11 module for the cards it supports. It looks like that is not the one you need. 
If your system had OpenSC you could find the default using the OpenSC tool:  `pkcs11-tool --help`

**Comment by jrgh0** [comment]
Hello again,

I'm only using opencryptoki with TPM by now, but I'm planning to also use cryptoauthlib with another chip on the SOM I use in my project. I've tested before cryptoauthlib with libp11 support and engine seems to work (I've not tested in deep yet).

I know libp11 is only a wraper for PKCS#11. If I've read well, documentation says that from 0.4.10 version it includes the engine for openssl. I figured libppkcs11.so is related to libp11 and pkcs11.so is related to the engine.

As the openssl engine from the opencryptoki library gave me an error and the project isn't very active, I was suggested to use libp11 engine by one of the maintainers of opencryptoki project.

So now, my question is, does the libp11 engine work with PKCS#11 TPM tokens from opencryptoki? If yes, what kind of error I'm facing?

Thank you & kind regards
J.R.
